### DISCUSSION TITLE
loadbalance切换节点时，可能会导致访问的网站速度变得很慢

### DISCUSSION THREAD

**Discussion by dev4u** [shadowsocks/shadowsocks-rust#757] [General]
我有两台服务器，A美国、B新加坡，已经配置负载均衡。

开始，我通过A访问youtube，一切很正常，整个访问链路：客户端->服务器A->美国的youtube cdn站点。
因为网络原因，sslocal切换到了服务器B。这个时候访问的链路变成：客户端->服务器B->美国的youtube cdn站点。
很明显，后者的访问链路虽然能通，但是并不是最优的。

导致这个现象，我估计是因为sslocal把解析到的网站地址缓存在客户端，当客户端被切换另一台服务器的时候，之前解析到的ip地址缓存没有清空，仍用于后续的服务器创建链接用。

能不能优化一下这块缓存逻辑？

**Comment by zonyitoo** [comment]
You just make an assumption and then ask for a solution...

First of all, as you said, `sslocal` detected some network problems to server A, so it switches to server B. Although the access path through B is not the optimal, but `sslocal` will then switch back to A if the network between them is recovered.

I don't know what exactly setup you were using. `sslocal` won't do any cache about target addresses.

**Reply by dev4u** [reply]
我请教一下，sslocal解析完一个域名，在ttl有效期内服务器地址发生切换，用户再次访问这个域名时，sslocal会重新发起域名解析吗？

**Reply by database64128** [reply]
`sslocal` only resolves your shadowsocks server’s domain name. Depending on your setup, it’s either your browser, system resolver, or `ssserver`’s job to lookup the domains of the website you are visiting.

**Reply by dev4u** [reply]
我懂你的意思了，谢谢指教。

**Comment by database64128** [comment]
If you open F12 and switch to the network tab while watching a video on YouTube, you'll notice that query parameters to YouTube's video CDN include your public IP address. When your actual public IP changes (`sslocal` switches server), you'll start to see failed requests with error code 403. I believe this is the type of "slowdowns" you are seeing in your observations. After a while of failed attempts, the player self-refreshes and starts to use your new public IP, and content starts to come through again.

**Reply by dev4u** [reply]
是的。
虽然访问的链接带了我的public ip，但访问的google服务仍然是美国节点。实际上访问新加坡节点才是更优的，这在域名ttl有效期内，会一直保持从新加坡访问美国节点的情况。

**Reply by database64128** [reply]
I don’t think it has anything to do with geolocation. It’s just failed attempts (403s) from IP change.

**Reply by dev4u** [reply]
> I don’t think it has anything to do with geolocation. It’s just failed attempts (403s) from IP change.

或许我能做的是修改返回更短的ttl长度。
